src/main/java/htsjdk/samtools/BamFileIoUtils.java

```diff
@@ -73,20 +96,21 @@ public static void reheaderBamFile(final SAMFileHeader samFileHeader, final File
      * @param skipHeader     If true, the header of the input file will not be copied to the output stream
      * @param skipTerminator If true, the terminator block of the input file will not be written to the output stream
      */
-    public static void blockCopyBamFile(final File inputFile, final OutputStream outputStream, final boolean skipHeader, final boolean skipTerminator) {
-        FileInputStream in = null;
-        try {
-            in = new FileInputStream(inputFile);
-
+    public static void blockCopyBamFile(final Path inputFile, final OutputStream outputStream, final boolean skipHeader, final boolean skipTerminator) {
+        // tsato: use CountingInputStream to replace FileInputStream. The latter has .getChannel().getPosition
+        // The regular InputStream from Files.newInputStream() doesn't have it, but I might be able to do so by creating Channel object and avoid using CountingInputStream
+        try (final CountingInputStream in = new CountingInputStream(Files.newInputStream(inputFile))){
+        // try (final InputStream in = Files.newInputStream(inputFile)){
             // a) It's good to check that the end of the file is valid and b) we need to know if there's a terminator block and not copy it if skipTerminator is true
             final BlockCompressedInputStream.FileTermination term = BlockCompressedInputStream.checkTermination(inputFile);
             if (term == BlockCompressedInputStream.FileTermination.DEFECTIVE)
-                throw new SAMException(inputFile.getAbsolutePath() + " does not have a valid GZIP block at the end of the file.");
+                throw new SAMException(inputFile.toUri() + " does not have a valid GZIP block at the end of the file.");
 
-            if (skipHeader) {
-                final long vOffsetOfFirstRecord = SAMUtils.findVirtualOffsetOfFirstRecordInBam(inputFile);
-                final BlockCompressedInputStream blockIn = new BlockCompressedInputStream(inputFile);
-                blockIn.seek(vOffsetOfFirstRecord);
+            if (skipHeader) { // tsato: this method assumes bam file...should I test cram?
+                final SeekablePathStream seekablePathStream = new SeekablePathStream(inputFile);
+                final long vOffsetOfFirstRecord = SAMUtils.findVirtualOffsetOfFirstRecordInBam(seekablePathStream);
+                final BlockCompressedInputStream blockIn = new BlockCompressedInputStream(seekablePathStream);
+                blockIn.seek(vOffsetOfFirstRecord); // This is why we give the BlockCompressedInputStream a SeekablePathStream rather than the regular InputStream
```

Comment:
Are you concerned that I am passing the same `seekablePathStream` object to `SAMUtils.findVirtualOffsetOfFirstRecordInBam()` and then to `new BlockCompressedInputStream()`, rather than creating a new Stream object for BlockCompressedInputStream?

Reply:
Yes.  That's exactly what I'm worried about.  I don't SEE a problem with it, but I'm always afraid there is some weird stream state being messed with in a way it shouldn't.  If it opened a new stream before I'd just do it again now.  (unless we want to make the effort to go through and actually understand everything which I didn't want to do...)

Reply:
done